added verb for filters and minor improvements

jmoriau opened the pull request:
I closed #96 and made this one because I forgot to make my changes on a separate branch.
Anyway here's the changelog:

param_parser.cr:
 - `@route` parameter was never unused.

context.cr:
 - changed `ParamParser.new` accordingly
 - added new `route_defined?` method, will be useful for other middlewares, also it’s lazily evaluated only once, so it can be called many times without any performances penalty.

route.cr:
 - both class methods didn’t really belong here in the first place, moved them to request.cr

request.cr: 
 - there’s no need to call `check_for_override!` it’s now lazily called when you access `.override_method` on `request`
 - calling many times `override_method` does not result in many wasted checks anymore (other middlewares will want to know the `override_method` before the `RouteHandler` checks it). Before it would have made the check happen twice, therefore wasting the second check. I realized this when writing the `before_all` and needed to know the `override_method` but didn’t want this to be called again later for nothing.

route_handler.cr:
 - removed the call for `check_override` since it’s no longer needed.
 - added `lookup_route` because I need to call it from a context from a middleware, this fetches the route from the tree.

filters.cr:
 - all the verbs are available including `all` (before_all, before_get etc...)
 - `Filter` middleware is now lazily instantiated and added to the handlers as soon as a call to `after_X` or `before_X` is made. No need to call `add_filters` anymore.
 - `add_filters` removed
 - `add` was renamed `add_route_filter` to be a bit more explicit.
 - the call order of the filters is before_all -> before_x -> X -> after_x -> after_all
 - added comments before each method to make the intent even clearer.
 - added specs for all the new methods.

Btw there is a new crystal 0.12 warning `embed_ecr` is deprecated, use `ECR.embed``
PS: I'll write the gh-pages documentation tomorrow I'm too tired to do it tonight :sob:

sdogruyol commented:
Wow. This one packs a lot of changes. I need to review this first :smile:

jmoriau commented:
Sure take your time :smile:

f commented:
🤔

waterlink commented on src/kemal/context.cr:
```diff
@@ -3,12 +3,18 @@
     end
+
+    def route_defined?
+      @route_defined = Kemal::RouteHandler::INSTANCE.lookup_route(@request.override_method as String, @request.path) if @route_defined == nil
```
This line looks too long. Maybe it makes sense to split it like that:

```crystal
@route_defined = Kemal::RouteHandler::INSTANCE.lookup_route(@request.override_method as String, @request.path)
```

waterlink replied:
This line looks too long. Maybe it makes sense to split it like that:

```crystal
if @route_defined == nil
  @route_defined = Kemal::RouteHandler::INSTANCE
    .lookup_route(@request.override_method as String, @request.path)
end
```

jmoriau replied:
it's a cosmetic change I don't mind :smile:

jmoriau commented:
I'm off to sleep I'm exhausted I'll check every suggestion you might have tomorrow :smile:

waterlink commented on src/kemal/middleware/filters.cr:
```diff
@@ -2,47 +2,72 @@ module Kemal::Middleware
-      lookup = @tree.find radix_path(type, context.request.path)
+    # This shouldn't be called directly, it's not private because I need to call it for testing purpose since I can't call the macros in the spec.
+    # It adds the block for the corresponding verb/path/type combination to the tree.
+    def add_route_filter(verb, path, type, &block : HTTP::Server::Context -> _)
```
Since it is not public API, but it can't be called from user's perspective, it makes sense to put `# :nodoc:` comment, so that it does not show up in a documentation.

Additionally, I like to name such methods with leading `_`, but that is just personal preference/opinion.

jmoriau replied:
Oh sure but I didn't know if the `_` was allowed as the first char of a method :smile: I'll change that tomorrow :+1:

waterlink commented:
Left some minor comments. Looks good to me otherwise.

jmoriau commented:
While trying to sleep I realized there is a small issue, I'll fix it tomorrow.

jmoriau commented:
This was preventing me from sleeping :sob:
Anyway I found the issue, which also led me to discover that the specs weren't run with the same `RouteHandler::INSTANCE` as normal `Kemal` app but instead they were using `RouteHandler.new` (obviously because it allowed to easily clear all the routes).
But this was making my change fail because `RouteHandler::INSTANCE` is used to lookup the route.
So I changed `RouteHandler.new` to `RouteHandler::INSTANCE` in the specs, and added `Kemal::RouteHandler::INSTANCE.tree = Radix::Tree.new` to get a new route tree for each spec.
I also made the changes @waterlink suggested, everything should be fine now and I'm dead tired :sleeping:  :zzz:

f commented:
It seems very clean to me. But it's a huge change and @sdogruyol should dig it :)

sdogruyol commented on src/kemal/middleware/filters.cr:
```diff
@@ -55,22 +80,15 @@ module Kemal::Middleware
-  filter.add :after, path, options, &block
-end
+ALL_METHODS = %w(get post put patch delete all)
+{% for type in ["before", "after"]%}
```
Let's move this macro to `dsl.cr`.

sdogruyol replied:
Or never mind. It's ok here :smile:

jmoriau replied:
I put it there first but then I thought since this is a middleware and optional it would make more sense to keep it separate from the `Kemal` dsl.
Are you sure :smile: ?

jmoriau replied:
Hahaha cool then :smile:

sdogruyol commented:
LGTM. Great job @jmoriau :100:

sdogruyol commented:
Just a small note here.

You can use the built-in `crystal tool format` to format the code :+1:

jmoriau commented:
Ah never used that, do you want me to do it and submit another PR ?

sdogruyol commented:
I've already done that :smile: Just a reminder for future PRs :+1:

jmoriau commented:
Ah awesome then :smile: thanks
